config/config.go

```diff
@@ -666,3 +697,17 @@ func (c *RemoteReadConfig) UnmarshalYAML(unmarshal func(interface{}) error) erro
 	// Thus we just do its validation here.
 	return c.HTTPClientConfig.Validate()
 }
+
+// ToHash generates a hash of the config struct minus it's name, as we consider configs
```

Comment:
Why is name not included? Wouldn't we want to restart the component so the metrics pick up the new name?

Reply:
Hmm, yeah I need to change this a bit. The idea is that two remote write configs that are identical except for their names should also be considered duplicates.

Reply:
@csmarchbanks PTAL, still think we might be able to simplify the hashing/comparison but my brain is dead atm. The current check works.

Reply:
Why do you want two remote write configs that are identical to be disallowed? I guess it doesn't make much sense to run that way, but moving Name logic to different places is extra complexity.

Reply:
Minimize the chance for users to end up with bad config. If we allow duplicates, but with different names, it means both queues would be writing to the same endpoint with the exact same samples. At the very least they'd end up with lots of write errors in their logs.

Needing both hashing functions is unfortunate, but to me it's better than maintaining a second version of the config struct that just removes the name field, and making one generic hash function. I could rename this function to `HashWithoutName` so it's clearer that they're different functions?

Reply:
:+1: Nit: `ToHash` -> `Hash` to be consistent with labels and such.

Reply:
> The idea is that two remote write configs that are identical except for their names should also be considered duplicates.

We don't consider that an issue for scrape configs, so I don't think we should consider it a problem here.

Reply:
I think what I mentioned above is reason enough to consider duplicates an issue for remote write even though they may not be for scrape.

Reply:
That's pretty much exactly what'd also happen with scrape configs, unless you know what you're up to.

Reply:
> We don't consider that an issue for scrape configs

That's not quite correct.

For one, we enforce a different job name for each scrape config, and by default, the job name becomes the `job` label for the scraped metric, so there isn't really a problem.

If, however, the config relabels the target labels so that they become identical, we _literally_ consider that an issue, namely [issue #5136](https://github.com/prometheus/prometheus/issues/5136).

I remote write config that differs in nothing but its name will create problems by definition, while a scrape config that differs in nothing but its name can very well be fine. That's a huge difference.

Reply:
> we literally consider that an issue, namely issue #5136.

That is a GIGO issue for the user, but not a problem for Prometheus.

Reply:
> That is a GIGO issue for the user, but not a problem for Prometheus.

I disagree. Of course, you are entitled to have your opinion. But please don't say "we" whan you mean "I".

I'm not sure if your statement is only about #5136.

I think it makes sense to reject duplicate remote-write configs with a different name. Do we have consensus on that?

Reply:
Every other place that it's been proposed to add something like that it's turned out not to be what we ended up doing, as either there's subtle cases where it's valid or it's a lot more complicated than it seems to catch all the corner cases. For example this could be relied upon for loadtesting, or there could be some weird load balancer involved.

I don't think we should start rejecting semantically valid configurations merely because they aren't necessarily sane.

#5136 is also about something subtly different, two targets with the same identity but not necessarily the same configuration.

Reply:
> Every other place that it's been proposed to add something like that it's turned out not to be what we ended up doing, as either there's subtle cases where it's valid or it's a lot more complicated than it seems to catch all the corner cases. For example this could be relied upon for loadtesting, or there could be some weird load balancer involved.

I leave it to the remote-write experts if they see anything problematic here. I don't see a corner case here is the configs are plainly identical. (We are not trying to catch cases where somebody happens to create different relabeling rules that happen to create the same output.) Loadtesting with all the samples duplicated would be problematic (I would loadtest the error detection), and I don't see how a loadbalancer could get involved in a reasonable way.

> I don't think we should start rejecting semantically valid configurations merely because they aren't necessarily sane.

Semantics is for us to define. I'd define identical configs as semantically invalid.

@cstyan @csmarchbanks what do you think?

Reply:
> Loadtesting with all the samples duplicated would be problematic

That depends on what you're loadtesting, it doesn't have to be the storage itself. It might be something like how well you deal with duplicate samples.

> I don't see how a loadbalancer could get involved in a reasonable way.

I said weird :)

> I'd define identical configs as semantically invalid.

I see no semantic issues here, nothing about it breaks our data, querying, or remote storage model. It works today and causes no problems for the Prometheus server. It may not have any "normal" use cases or be efficient, but I don't think we should be 2nd guessing what users put in their config file.

Reply:
> It might be something like how well you deal with duplicate samples.

This line of arguments would require the removal of all kind of correctness check. Should I now allow duplicates in expositions by client_golang because somebody might want to test how well a consumer deals with duplicate samples?

> It works today and causes no problems for the Prometheus server.

Again, with this line of argument, we could justify all kind of insane output of the Prometheus server.

I expect remote-write config duplication to happen by accident, and I want to protect my users from it. While accidental config duplication is likely to happen, valid use cases for duplicate config is all but speculation at this point. The course of action appears clear to me from that.

Reply:
I agree, I think it's semantically incorrect to accept identical configurations in the case of remote write when duplicates have implications for the downstream remote storage.

Reply:
> in the case of remote write when duplicates have implications for the downstream remote storage.

That's something that we get to define rather than the remote storages, and as we're going for at-least-once delivery remote storages already have to handle this semantically. Efficiency is not the same as semantics.

Reply:
I think that providing some sanity checking of configuration would be nice, but I also think it will be weird that two configs with a different backoff is counted as "valid" but two configs with different names are rejected. As far as this PR goes, my vote would be that the name is included in any hash, and required to be unique due to the metrics we export. That would allow a user to just change the name of the queue in case they want duplicate samples being sent, but would guard against simple copy-paste without changing anything mistakes.

I am not opposed to the current implementation, just a preference to include the name in the hash. Finally, perhaps it would be good to open a separate issue/PR for sanity checking effectively duplicate configs more thoroughly?